src/reducers/measurements/types.ts

```diff
@@ -0,0 +1,100 @@
+import { measurementIdSymbol } from "../../util/globals";
+
+// -- Measurements JSON types -- //
+
+/**
+ * Measurements are allowed to have arbitrary metadata.
+ * Matching types allowed in Augur's measurements schema
+ * <https://github.com/nextstrain/augur/blob/3f72c40897a80132099729d5d00a6718e76e0e9e/augur/data/schema-measurements.json#L152>
+ */
+type MeasurementMetadata = string | number | boolean
+
+export const measurementsDisplayValues = ["raw", "mean"];
+export type MeasurementsDisplay = (typeof measurementsDisplayValues)[number]
+export const isMeasurementsDisplay = (x: any): x is MeasurementsDisplay => measurementsDisplayValues.includes(x);
+
+interface JsonMeasurement {
+  readonly strain: string
+  readonly value: number
+  readonly [key: string]: MeasurementMetadata
+}
+
+interface JsonCollectionDisplayDefaults {
+  readonly group_by?: string
+  readonly measurements_display?: MeasurementsDisplay
+  readonly show_overall_mean?: boolean
+  readonly show_threshold?: boolean
+}
+
+interface JsonCollectionField {
+  readonly key: string
+  readonly title?: string
+}
+
+interface JsonCollectionGrouping {
+  readonly key: string
+  readonly order?: MeasurementMetadata[]
+}
+
+export interface JsonCollection {
+  readonly display_defaults?: JsonCollectionDisplayDefaults
+  readonly fields?: JsonCollectionField[]
+  readonly filters?: string[]
+  readonly groupings: JsonCollectionGrouping[]
+  readonly key: string
+  readonly measurements: JsonMeasurement[]
+  readonly threshold?: number
+  readonly thresholds?: number[]
+  readonly title?: string
+  readonly x_axis_label: string
+}
+
+export interface MeasurementsJson {
+  readonly collections: JsonCollection[]
+  readonly default_collection?: string
+}
+
+// -- Measurements state types -- //
+
+export interface Measurement {
+  [measurementIdSymbol]: number
+  strain: string
+  value: number
+  [key: string]: string | number
```

Comment:
I was confused why the `typeof` check here was necessary despite 33afa42a00bc0c07219689ab93ddd6ce73e93063:

https://github.com/nextstrain/auspice/blob/891488a7ba881cb054dfa2646268f6e4cbad9f38/src/components/measurements/index.tsx#L143

Should this type be `[key: string]: string` without `number`?

Reply:
Oh true, I should've tried first. I think this [intersection type](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#intersection-types) works:

```ts
export type Measurement = {
  [measurementIdSymbol]: number
  strain: string
  value: number
} & {
  [key: string]: string
}
```

Reply:
OHHH, interesting! Seems weird that it works, but I'll go with it (with some comments).

Reply:
Hah, the intersection type doesn't show errors for the type declaration but runs into errors when I tried to implement https://github.com/nextstrain/auspice/pull/1910#discussion_r1848965620. 

```
src/reducers/measurements/types.ts:76:5 - error TS2322: Type '{ [measurementIdSymbol]: number; strain: string; value: number; }' is not assignable to type 'Measurement'.
  Type '{ [measurementIdSymbol]: number; strain: string; value: number; }' is not assignable to type '{ [key: string]: string; }'.
    Property 'value' is incompatible with index signature.
      Type 'number' is not assignable to type 'string'.

 76     return {
        ~~~~~~~~
 77        ...x,
    ~~~~~~~~~~~~
... 
 80        value: x.value,
    ~~~~~~~~~~~~~~~~~~~~~~
 81     }
    ~~~~~
```

I'm going to keep `[key: string]: string | number` for now and will revisit later.

Reply:
Will revisit as part of https://github.com/nextstrain/auspice/issues/1918